app/scripts/metamask-controller.js

```diff
@@ -6324,6 +6325,168 @@ export default class MetamaskController extends EventEmitter {
+      return;
+    }
+
+    if (shouldShowApproval) {
+      try {
+        const approvalResponse =
+          await this.approvalController.addAndShowApprovalRequest({
+            origin,
+            type: HYPERLIQUID_APPROVAL_TYPE,
+            requestData: { selectedAddress: permittedAccount },
+          });
+
+        if (approvalResponse?.approved) {
+          this._handleHyperliquidApprovedAccount(
+            permittedAccount,
+            permittedAccounts,
+            declinedAccounts,
+          );
+          await this._handleHyperliquidReferralRedirect(
+            tabId,
+            HYPERLIQUID_ORIGIN,
+            permittedAccount,
+          );
+        } else {
+          this.preferencesController.addReferralDeclinedAccount(
+            permittedAccount,
+          );
+        }
+      } catch (error) {
+        // We expect to get here if user switches account without approving
+      }
+    }
+
+    if (shouldRedirect) {
+      await this._handleHyperliquidReferralRedirect(
+        tabId,
+        HYPERLIQUID_ORIGIN,
+        permittedAccount,
+      );
+    }
+  }
+
+  /**
+   * Handles redirection to the Hyperliquid referral page.
+   *
+   * @param {number} tabId - The browser tab ID to update.
+   * @param {string} hyperliquidOrigin - The Hyperliquid origin URL.
+   * @param {string} permittedAccount - The permitted account.
+   */
+  async _handleHyperliquidReferralRedirect(
+    tabId,
+    hyperliquidOrigin,
+    permittedAccount,
+  ) {
+    await this._updateHyperliquidReferralUrl(tabId, hyperliquidOrigin);
+    // Mark this account as having been shown the Hyperliquid referral page
+    this.preferencesController.addReferralPassedAccount(permittedAccount);
+  }
+
+  /**
+   * Handles referral states for permitted accounts after user approval.
+   *
+   * @param {string} permittedAccount - The permitted account.
+   * @param {string[]} permittedAccounts - The permitted accounts.
+   * @param {string[]} declinedAccounts - The previously declined permitted accounts.
+   */
+  _handleHyperliquidApprovedAccount(
+    permittedAccount,
+    permittedAccounts,
+    declinedAccounts,
+  ) {
+    if (declinedAccounts.length === 0) {
+      // If there are no previously declined permitted accounts then
+      // we approve all permitted accounts so that the user is not
+      // shown the approval screen unnecessarily when switching
+      this.preferencesController.setAllAccountsReferralApproved(
+        permittedAccounts,
+      );
+    } else {
+      this.preferencesController.addReferralApprovedAccount(permittedAccount);
+      // If there are any previously declined accounts then
+      // we do not approve them, but instead remove them from the declined list
+      // so they have the option to participate again in future
+      permittedAccounts.slice(1).forEach((account) => {
+        if (declinedAccounts.includes(account)) {
+          this.preferencesController.removeReferralDeclinedAccount(account);
```

Comment:
This is odd to me. Does approving one account mean that you are open to approving accounts you have previously declined?

Reply:
Yes that is the logic we are following. If there are no declined accounts we approve all the connected accounts. If there is a declined account but a user approves a different account, we remove the declined status from the account so next time they connect they are shown the confirmation again. The reasoning behind these decisions is optimise the chance of user accepting the referral whilst also trying to reduce friction in showing the confirmation screen repeatedly once it's accepted.

Reply:
I don't totally agree with this UX decision. The problem is that there is no way to say "no, please don't ask me again". I can see the argument that if a user approved a referral once that maybe they'd be open to approving accounts they previously rejected, but it still feels odd